api/src/test/java/org/openmrs/api/OrderSetServiceTest.java

```diff
@@ -0,0 +1,156 @@
+		}
+		if (null == orderService) {
+			orderService = Context.getOrderService();
+		}
+		if (null == conceptService) {
+			conceptService = Context.getConceptService();
+		}
+	}
+	
+	@After
+	public void revertToDefaultLocale() throws Exception {
+		Context.setLocale(Locale.US);
+	}
+	
+	/**
+	 * Updates the search index to clean up after each test.
+	 *
+	 * @see org.openmrs.test.BaseContextSensitiveTest#updateSearchIndex()
+	 */
+	@Before
+	@Override
+	public void updateSearchIndex() {
+		super.updateSearchIndex();
+	}
+	
+	/**
+	 * Updates the search index after executing each dataset.
+	 *
+	 * @see org.openmrs.test.BaseContextSensitiveTest#executeDataSet(org.dbunit.dataset.IDataSet)
+	 */
+	@Override
+	public void executeDataSet(IDataSet dataset) throws Exception {
+		super.executeDataSet(dataset);
+		updateSearchIndex();
+	}
+	
+	@Test
+	@Verifies(value = "should save OrderSet", method = "save(OrderSet)")
+	public void shouldSaveOrderSet() throws Exception {
+		executeDataSet(ORDER_SET);
+		OrderSet orderSet = orderSetBuilder(false, false);
+		OrderSet orderSetObj = orderSetService.save(orderSet);
+		assertNotNull("OrderSet has a order_set_id", orderSetObj.getId().intValue());
+	}
+	
+	@Test
+	@Verifies(value = "should update existing OrderSet", method = "save(OrderSet)")
+	public void shouldSaveAndUpdateOrderSet() throws Exception {
+		executeDataSet(ORDER_SET);
+		OrderSet orderSet = orderSetBuilder(false, false);
+		orderSet.setRetired(true);
+		OrderSet orderSetObj = orderSetService.save(orderSet);
+		assertEquals("OrderSet should contain orderSetmembers", 1, orderSetObj.getOrderSetMembers().size());
+		
+		orderSetObj.setOperator(OrderSet.Operator.ONE);
+		orderSetObj.setDescription("Test Order Set Description Updated");
+		orderSetObj.setOrderSetMembers(new HashSet<OrderSetMember>());
+		orderSetObj = orderSetService.save(orderSetObj);
+		assertNotNull("OrderSet has a order_set_id", orderSetObj.getId().intValue());
+		assertEquals("OrderSet has updated description", "Test Order Set Description Updated", orderSetObj.getDescription());
+		assertEquals("OrderSet has updated operator", "ONE", orderSetObj.getOperator().toString());
+		assertEquals("OrderSet should not contain orderSetmembers", 0, orderSetObj.getOrderSetMembers().size());
+		
+	}
+	
+	@Test
+	public void shouldRetrieveAllOrderSetsWithOrderSetMember() throws Exception {
+		executeDataSet(ORDER_SET);
+		List<OrderSet> olderNumberOfOrderSets = orderSetService.getOrderSets(false);
+		OrderSet orderSet = orderSetBuilder(false, false);
+		OrderSet orderSetObj = orderSetService.save(orderSet);
+		assertEquals("OrderSet should contain orderSetmembers", 1, orderSetObj.getOrderSetMembers().size());
+		
+		List<OrderSet> orderSets = orderSetService.getOrderSets(false);
+		assertEquals("OrderSet should be saved in DB", olderNumberOfOrderSets.size() + 1, orderSets.size());
+		assertEquals("OrderSet should not contain voided orderSetMembers", 1, orderSetObj.getOrderSetMembers().size());
```

Comment:
This seems a bit confusing, for this to make sense i would assume the test data should be set up in such a way that you have some voided and none retired members, when "OrderSet should contain orderSetmembers" and "OrderSet should not contain voided orderSetmembers" both are checking for the same count it makes it unclear and shouldn't be 'retired' and not 'voided'. Also getSetMembers needs to be overloaded to have another variation that takes a boolean flag that indicates if retired members should be included. I would also assume passing false to OrderService.getOrderSets should not automatically propagate to the set members of the OrderSets i.e saying that exclude retired OrderSets doesn't translate to also excluding retired set members for each of the returned sets